File: examples/agent-client/ts/pay-and-fetch.ts
/**
 * BYTE x402 — agent-pays demo (TypeScript, PRIMARY)
 * =================================================
 * An autonomous agent that GETs a paid BYTE feed, lets the x402 client
 * auto-handle the HTTP 402 by signing a gasless USDC payment on Base, retries,
 * and prints the feed data + the on-chain settlement tx hash + a Basescan link.
 *
 * Matches the gateway's SDK exactly: @x402/core + @x402/evm v2.13.x, viem.
 * No high-level `wrapFetchWithPayment` exists in this v2 SDK — we compose the
 * documented primitives: x402Client (with ExactEvmScheme) -> x402HTTPClient ->
 * a small fetch loop. The wallet only SIGNS an EIP-3009 authorization; the
 * facilitator broadcasts + pays gas, so the agent needs USDC but no ETH.
 *
 * Run:
 *   npm install
 *   AGENT_PRIVATE_KEY=0x... npm start                 # defaults to x402-pulse on Base Sepolia
 *   AGENT_PRIVATE_KEY=0x... GATEWAY_URL=https://x402.payperbyte.io \
 *     FEED_PATH=/feeds/x402-pulse NETWORK=base-mainnet npm start
 *
 * Env:
 *   AGENT_PRIVATE_KEY  (REQUIRED) funded test/agent wallet private key. NEVER hardcode.
 *   GATEWAY_URL        gateway base URL (default http://127.0.0.1:3402)
 *   FEED_PATH          feed endpoint to buy (default /feeds/x402-pulse)
 *   NETWORK            "base-sepolia" (default) | "base-mainnet" — picks chain + Basescan host
 *   RPC_URL            optional override for the Base RPC (defaults to viem public RPC)
 */

import { privateKeyToAccount } from "viem/accounts";
import { createPublicClient, http } from "viem";
import { base, baseSepolia } from "viem/chains";
import { x402Client, x402HTTPClient } from "@x402/core/client";
import { registerExactEvmScheme } from "@x402/evm/exact/client";
import { toClientEvmSigner } from "@x402/evm";

// ── Config from env ────────────────────────────────────────────────────────
const PRIVATE_KEY = process.env.AGENT_PRIVATE_KEY as `0x${string}` | undefined;
const GATEWAY_URL = (process.env.GATEWAY_URL ?? "http://127.0.0.1:3402").replace(/\/$/, "");
const FEED_PATH = process.env.FEED_PATH ?? "/feeds/x402-pulse";
const NETWORK = process.env.NETWORK ?? "base-sepolia";

if (!PRIVATE_KEY) {
  console.error("ERROR: set AGENT_PRIVATE_KEY (a funded Base test wallet). Never hardcode it.");
  process.exit(1);
}
if (!/^0x[0-9a-fA-F]{64}$/.test(PRIVATE_KEY)) {
  console.error("ERROR: AGENT_PRIVATE_KEY must be a 0x-prefixed 32-byte hex private key.");
  process.exit(1);
}

const isMainnet = NETWORK === "base-mainnet";
const chain = isMainnet ? base : baseSepolia;
const basescan = isMainnet ? "https://basescan.org" : "https://sepolia.basescan.org";

// ── Wallet + signer ─────────────────────────────────────────────────────────
// The viem account satisfies the x402 ClientEvmSigner interface (address +
// signTypedData). We wrap it with a public client via toClientEvmSigner so the
// scheme can do optional on-chain reads (nonce/allowance) when needed.
const account = privateKeyToAccount(PRIVATE_KEY);
const publicClient = createPublicClient({
  chain,
  transport: http(process.env.RPC_URL),
});
const signer = toClientEvmSigner(account, publicClient);

// ── x402 client: exact-EVM scheme on Base, wrapped for HTTP ─────────────────
const core = registerExactEvmScheme(new x402Client(), { signer });
const x402 = new x402HTTPClient(core);

/**
 * One paid request. Sends the request; if the gateway answers 402, signs a
 * payment for the advertised requirements and retries with the payment header;
 * returns the parsed x402PaymentResult.
 */
async function payAndFetch(url: string, init: RequestInit = {}) {
  // 1) Unpaid attempt — discovers the 402 + payment requirements.
  const first = await fetch(url, init);
  if (first.status !== 402) {
    // Free route, error, or already paid — hand back as-is.
    return { result: await x402.processResponse(first), paid: false };
  }

  // 2) Parse the PaymentRequired (header on v2, body on v1) and build a payload.
  const paymentRequired = x402.getPaymentRequiredResponse(
    (name) => first.headers.get(name),
    await first.clone().json().catch(() => undefined),
  );
  const payload = await x402.createPaymentPayload(paymentRequired);
  const paymentHeaders = x402.encodePaymentSignatureHeader(payload);

  // 3) Retry the SAME request with the signed payment header attached.
  const paidResp = await fetch(url, {
    ...init,
    headers: { ...(init.headers ?? {}), ...paymentHeaders },
  });
  return { result: await x402.processResponse(paidResp), paid: true };
}

async function main() {
  const url = `${GATEWAY_URL}${FEED_PATH}`;
  console.log(`[agent] wallet      : ${account.address}`);
  console.log(`[agent] network     : ${NETWORK} (chainId ${chain.id})`);
  console.log(`[agent] buying      : GET ${url}`);

  const { result, paid } = await payAndFetch(url, { method: "GET" });

  switch (result.kind) {
    case "success": {
      const s = result.settleResponse;
      console.log(`\n[agent] PAID + SETTLED ${paid ? "" : "(was free)"}`.trimEnd());
      console.log(`[agent] settlement  : success=${s.success} payer=${s.payer ?? account.address}`);
      console.log(`[agent] tx hash     : ${s.transaction}`);
      console.log(`[agent] basescan    : ${basescan}/tx/${s.transaction}`);
      console.log(`\n[agent] feed data   :\n${JSON.stringify(result.body, null, 2)}`);
      break;
    }
    case "passthrough":
      console.log(`\n[agent] free / no payment required:\n${JSON.stringify(result.body, null, 2)}`);
      break;
    case "settle_failed":
      console.error(`\n[agent] PAYMENT SIGNED BUT SETTLE FAILED.`);
      console.error(`[agent] settleResponse: ${JSON.stringify(result.settleResponse, null, 2)}`);
      console.error(`[agent] body: ${JSON.stringify(result.body, null, 2)}`);
      process.exit(2);
      break;
    case "payment_required":
      console.error(`\n[agent] still 402 after paying — requirements:`);
      console.error(JSON.stringify(result.paymentRequired, null, 2));
      process.exit(2);
      break;
    case "error":
      console.error(`\n[agent] gateway error ${result.status}: ${JSON.stringify(result.body)}`);
      process.exit(2);
      break;
  }
}

main().catch((err) => {
  console.error(`[agent] fatal: ${err instanceof Error ? err.stack : err}`);
  process.exit(1);
});

File: examples/agent-client/python/pay_and_fetch.py
#!/usr/bin/env python3
"""
BYTE x402 - agent-pays demo (Python, SECONDARY)
===============================================
An autonomous agent that GETs a paid BYTE feed, lets the x402 requests client
auto-handle the HTTP 402 by signing a gasless USDC payment on Base, retries,
and prints the feed data + the on-chain settlement tx hash + a Basescan link.

Matches the gateway's protocol exactly: x402 v2 "exact" EVM scheme, EIP-3009.
The wallet only SIGNS an authorization; the facilitator broadcasts and pays gas,
so the agent needs USDC but no ETH.

Built on the official x402 python package (== the gateway's @x402 v2 line):
    x402.client.x402ClientSync
    x402.mechanisms.evm.exact.register.register_exact_evm_client
    x402.mechanisms.evm.signers.EthAccountSigner  (eth_account / web3)
    x402.http.clients.requests.x402_requests       (auto-paying requests.Session)

Install:
    python3 -m venv .venv && . .venv/bin/activate
    pip install -r requirements.txt

Run:
    AGENT_PRIVATE_KEY=0x... python pay_and_fetch.py            # x402-pulse, Base Sepolia
    AGENT_PRIVATE_KEY=0x... GATEWAY_URL=https://x402.payperbyte.io \
        FEED_PATH=/feeds/x402-pulse NETWORK=base-mainnet python pay_and_fetch.py

Env:
    AGENT_PRIVATE_KEY  (REQUIRED) funded test/agent wallet private key. NEVER hardcode.
    GATEWAY_URL        gateway base URL (default http://127.0.0.1:3402)
    FEED_PATH          feed endpoint to buy (default /feeds/x402-pulse)
    NETWORK            "base-sepolia" (default) | "base-mainnet" (picks Basescan host)
    RPC_URL            optional Base RPC (enables on-chain nonce/allowance reads)
"""
import json
import os
import re
import sys

import requests
from eth_account import Account

from x402.client import x402ClientSync
from x402.http.clients.requests import x402_requests
from x402.http.constants import PAYMENT_RESPONSE_HEADER, X_PAYMENT_RESPONSE_HEADER
from x402.http.utils import decode_payment_response_header
from x402.mechanisms.evm.exact.register import register_exact_evm_client
from x402.mechanisms.evm.signers import EthAccountSigner, EthAccountSignerWithRPC

# -- Config from env ---------------------------------------------------------
PRIVATE_KEY = os.environ.get("AGENT_PRIVATE_KEY")
GATEWAY_URL = os.environ.get("GATEWAY_URL", "http://127.0.0.1:3402").rstrip("/")
FEED_PATH = os.environ.get("FEED_PATH", "/feeds/x402-pulse")
NETWORK = os.environ.get("NETWORK", "base-sepolia")
RPC_URL = os.environ.get("RPC_URL")

if not PRIVATE_KEY:
    sys.exit("ERROR: set AGENT_PRIVATE_KEY (a funded Base test wallet). Never hardcode it.")
if not re.fullmatch(r"0x[0-9a-fA-F]{64}", PRIVATE_KEY):
    sys.exit("ERROR: AGENT_PRIVATE_KEY must be a 0x-prefixed 32-byte hex private key.")

is_mainnet = NETWORK == "base-mainnet"
basescan = "https://basescan.org" if is_mainnet else "https://sepolia.basescan.org"

# -- Wallet + signer ---------------------------------------------------------
# EthAccountSigner is sign-only (sufficient: gasless EIP-3009, facilitator
# broadcasts). EthAccountSignerWithRPC adds nonce/allowance reads when RPC_URL
# is set - harmless either way for the "exact" USDC path.
account = Account.from_key(PRIVATE_KEY)
signer = EthAccountSignerWithRPC(account, RPC_URL) if RPC_URL else EthAccountSigner(account)

# -- x402 client: exact-EVM scheme, wrapped into an auto-paying Session -------
client = x402ClientSync()
# No `networks=` arg -> registers the eip155:* wildcard, which covers Base
# mainnet (eip155:8453) and Base Sepolia (eip155:84532). The gateway's 402
# names the concrete network + USDC asset; the client pays whatever it asks.
register_exact_evm_client(client, signer)
session = x402_requests(client)


def extract_settlement(resp: requests.Response):
    """Decode the settlement SettleResponse from the response headers (v2 then v1)."""
    raw = resp.headers.get(PAYMENT_RESPONSE_HEADER) or resp.headers.get(X_PAYMENT_RESPONSE_HEADER)
    if not raw:
        return None
    try:
        return decode_payment_response_header(raw)
    except Exception as exc:  # noqa: BLE001
        print(f"[agent] warning: could not decode settlement header: {exc}", file=sys.stderr)
        return None


def main() -> None:
    url = f"{GATEWAY_URL}{FEED_PATH}"
    print(f"[agent] wallet      : {account.address}")
    print(f"[agent] network     : {NETWORK}")
    print(f"[agent] buying      : GET {url}")

    # The wrapped session transparently does: GET -> 402 -> sign USDC payment ->
    # retry with payment header -> return the final (paid) response.
    resp = session.get(url, timeout=60)

    if resp.status_code != 200:
        print(f"[agent] gateway returned {resp.status_code}: {resp.text}", file=sys.stderr)
        sys.exit(2)

    settle = extract_settlement(resp)
    if settle is not None:
        print("\n[agent] PAID + SETTLED")
        print(f"[agent] settlement  : success={settle.success} payer={settle.payer or account.address}")
        print(f"[agent] tx hash     : {settle.transaction}")
        print(f"[agent] basescan    : {basescan}/tx/{settle.transaction}")
    else:
        print("\n[agent] response had no settlement header (free route or already paid).")

    try:
        body = resp.json()
        print(f"\n[agent] feed data   :\n{json.dumps(body, indent=2)}")
    except ValueError:
        print(f"\n[agent] feed data (raw):\n{resp.text}")


if __name__ == "__main__":
    main()
